File: Dynamic-Programming/63.py
"""
63. 不同路径 II
中等
https://leetcode-cn.com/problems/unique-paths-ii/
"""

from typing import List

class Solution:
    # 方法一: 动态规划思路，到重点的路径等于左边的加上上面的两个点的路径和

    # 48 ms	    35.96%
    # 13.8 MB   14.29%
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        dp = [[0] * n for _ in range(m)]
        if obstacleGrid[0][0] == 1 or obstacleGrid[m - 1][n - 1]:
            return 0
        if m == n == 1:
            return 1
        dp[0][0] = 1
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    continue
                if obstacleGrid[i][j] == 0:
                    path = 0
                    if j - 1 >= 0:
                        if obstacleGrid[i][j - 1] == 0:
                            path += dp[i][j - 1]
                    if i - 1 >= 0:
                        if obstacleGrid[i - 1][j] == 0:
                            path += dp[i - 1][j]
                    dp[i][j] = path
        if m == 1:
            return dp[0][n - 1]
        if n == 1:
            return dp[m - 1][0]
        return dp[m - 1][n - 2] + dp[m - 2][n - 1]

if __name__ == "__main__":
    su = Solution()
    test = [
        [[0],[0]],
        [[0,0,0,0]],
        [
            [0,0,0,0,0],
            [0,1,0,1,0],
            [0,0,0,0,0]
        ],
        [[0,0,0],[0,1,0],[0,0,0]],
        [[0,0],[0,1]]
    ]
    for i in test:
        print(su.uniquePathsWithObstacles(i))



[[0,0,0],[0,1,0],[0,0,0]]
[[0,0,0,0,0],[0,1,0,1,0],[0,0,0,0,0]]
[[1]]
[[0]]
[[0,1,1],[0,0,0]]
[[0,0,1],[1,0,0],[0,1,0]]
[[0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0]]
[[0,0],[0,0]]
[[0],[0]]
[[0,0,0,0]]
[[0,0],[0,1]]

File: Leetcode-Swift/Leetcode/63.swift
//
//  63.swift
//  Leetcode-Swift
//
//  Created by devcz on 2022/5/21.
//

import Foundation

/*
 63. 不同路径 II (中等)
 https://leetcode.cn/problems/unique-paths-ii/
 */

class Solution_63 {
    func uniquePathsWithObstacles(_ obstacleGrid: [[Int]]) -> Int {
        if obstacleGrid.isEmpty || obstacleGrid[0].isEmpty {
            return 0
        }

        if obstacleGrid[0][0] == 1 {
            return 0
        }

        let rows = obstacleGrid.count
        let cols = obstacleGrid[0].count
        var dp = [[Int]](repeating: [Int](repeating: 0, count: cols), count: rows)
        dp[0][0] = 1

        for i in 1..<rows {
            if obstacleGrid[i][0] == 1 {
                break
            }
            dp[i][0] = 1
        }
        for j in 1..<cols {
            if obstacleGrid[0][j] == 1 {
                break
            }
            dp[0][j] = 1
        }

        for i in 1..<rows {
            for j in 1..<cols {
                if obstacleGrid[i][j] == 1 {
                    continue
                }

                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
            }
        }

        return dp[rows - 1][cols - 1]
    }
}
